tracer/src/Datadog.Trace/Sampling/SamplingRuleHelper.cs

```diff
@@ -0,0 +1,131 @@
+        List<KeyValuePair<string, Regex?>>? tagRegexes,
+        out bool timedOut)
+    {
+        timedOut = false;
+
+        if (span == null!)
+        {
+            return false;
+        }
+
+        try
+        {
+            // if a regex is null (not specified), it always matches.
+            // stop as soon as we find a non-match.
+            return IsMatch(serviceNameRegex, span.ServiceName) &&
+                   IsMatch(operationNameRegex, span.OperationName) &&
+                   IsMatch(resourceNameRegex, span.ResourceName) &&
+                   MatchSpanByTags(span, tagRegexes);
+        }
+        catch (RegexMatchTimeoutException e)
+        {
+            // flag rule so we don't try to use one of its regexes again
+            timedOut = true;
+
+            Log.Error(
+                e,
+                """Regex timed out when trying to match value "{Input}" against pattern "{Pattern}".""",
+                e.Input,
+                e.Pattern);
+
+            return false;
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static bool IsMatch(Regex? regex, string? input)
+    {
+        if (regex is null)
+        {
+            // if a regex is null (not specified), it always matches.
+            return true;
+        }
+
+        if (input is null)
+        {
+            return false;
+        }
+
+        return regex.Match(input).Success;
+    }
+
+    private static bool MatchSpanByTags(Span span, List<KeyValuePair<string, Regex?>>? tagRegexes)
+    {
+        if (tagRegexes is null || tagRegexes.Count == 0)
+        {
+            // if a regex is null (not specified), it always matches.
+            return true;
+        }
+
+        foreach (var pair in tagRegexes)
+        {
+            var tagName = pair.Key;
+            var tagRegex = pair.Value;
+
+            if (tagRegex is null)
+            {
+                // if a regex is null (not specified), it always matches.
+                continue;
+            }
+
+            var tagValue = GetSpanTag(span, tagName);
+
+            if (tagValue is null || !tagRegex.Match(tagValue).Success)
+            {
+                // stop as soon as we find a tag that isn't set or doesn't match
+                return false;
+            }
+        }
+
+        // all specified tags exist and matched
+        return true;
+    }
+
+    private static string? GetSpanTag(Span span, string tagName)
+    {
+        if (span.GetTag(tagName) is { } tagValue)
+        {
+            return tagValue;
+        }
+
+        // if the string tag doesn't exist, try to get it as a numeric tag...
+        if (span.GetMetric(tagName) is not { } numericTagValue)
+        {
+            return null;
+        }
+
+        // ...but only if it is an integer
+        var intValue = (int)numericTagValue;
+        return Math.Abs(intValue - numericTagValue) < 0.0001 ? intValue.ToString(CultureInfo.InvariantCulture) : null;
```

Comment:
How come we have to test the absolute value? Is a simple int conversion insufficient?

Reply:
I want to determine if the tag value is an integer, but it is stored as a `double`. If you convert it to `int` and try to compare them with `(intValue == doubleValue)`, Rider recommends using `Math.Abs(value1 - value2) < tolerance)` instead.

> Using the `==`/`!=` operators to compare floating-point numbers is, generally, a bad idea. Floating point values are inherently inaccurate, and comparing them for exact equality is almost never the desired semantics. 

https://www.jetbrains.com/help/rider/CompareOfFloatsByEqualityOperator.html